Library/Homebrew/test/test_formula.rb

```diff
@@ -591,23 +639,97 @@ def test_mixed_taps_outdated_version_installed
     extra_outdated_prefix = HOMEBREW_CELLAR/"#{f.name}/1.0"
 
     setup_tab_for_prefix(outdated_prefix)
-    setup_tab_for_prefix(extra_outdated_prefix, "homebrew/core")
+    setup_tab_for_prefix(extra_outdated_prefix, :tap => "homebrew/core")
+    reset_outdated_versions
 
     refute_predicate f.outdated_versions, :empty?
 
-    setup_tab_for_prefix(outdated_prefix, "user/repo")
+    setup_tab_for_prefix(outdated_prefix, :tap => "user/repo")
+    reset_outdated_versions
 
     refute_predicate f.outdated_versions, :empty?
   end
 
   def test_same_version_tap_installed
     f.instance_variable_set(:@tap, CoreTap.instance)
-    setup_tab_for_prefix(same_prefix, "homebrew/core")
+    setup_tab_for_prefix(same_prefix, :tap => "homebrew/core")
+
+    assert_predicate f.outdated_versions, :empty?
+
+    setup_tab_for_prefix(same_prefix, :tap => "user/repo")
+    reset_outdated_versions
 
     assert_predicate f.outdated_versions, :empty?
+  end
+
+  def test_outdated_installed_head_less_than_stable
+    tab = setup_tab_for_prefix(head_prefix, :versions => { "stable" => "1.0" })
+    refute_predicate f.outdated_versions, :empty?
 
-    setup_tab_for_prefix(same_prefix, "user/repo")
+    # Tab.for_keg(head_prefix) will be fetched from CACHE but we write it anyway
+    tab.source["versions"] = { "stable" => f.version.to_s }
+    tab.write
+    reset_outdated_versions
 
     assert_predicate f.outdated_versions, :empty?
   end
+
+  def test_outdated_fetch_head
+    outdated_stable_prefix = HOMEBREW_CELLAR.join("testball/1.0")
+    head_prefix_a = HOMEBREW_CELLAR.join("testball/HEAD")
+    head_prefix_b = HOMEBREW_CELLAR.join("testball/HEAD-aaaaaaa_1")
+    head_prefix_c = HOMEBREW_CELLAR.join("testball/HEAD-56589463e5d9942a19d8e80975034ebf8886bba4")
+
+    setup_tab_for_prefix(outdated_stable_prefix)
+    tab_a = setup_tab_for_prefix(head_prefix_a, :versions => { "stable" => "1.0" })
+    setup_tab_for_prefix(head_prefix_b)
+
+    initial_env = ENV.to_hash
+    testball_repo = HOMEBREW_PREFIX.join("testball_repo")
+    testball_repo.mkdir
+
+    @f = formula("testball") do
+      url "foo"
+      version "2.10"
+      head "file://#{testball_repo}", :using => :git
+    end
+
+    %w[AUTHOR COMMITTER].each do |role|
+      ENV["GIT_#{role}_NAME"] = "brew tests"
+      ENV["GIT_#{role}_EMAIL"] = "brew-tests@localhost"
+      ENV["GIT_#{role}_DATE"] = "Thu May 21 00:04:11 2009 +0100"
+    end
+
+    testball_repo.cd do
+      FileUtils.touch "LICENSE"
+      shutup do
+        system "git", "init"
+        system "git", "add", "--all"
+        system "git", "commit", "-m", "Initial commit"
+      end
+    end
+
+    refute_predicate f.outdated_versions(:fetch_head => true), :empty?
+
+    tab_a["versions"] = { "stable" => f.version }
```

Comment:
The test failure is actually happening here. (You cannot fully trust the line numbers in the test bot output because your PR is being merged on top of an up-to-date `master`, so things might move a little bit.) The problem is that `OpenStruct` in Ruby 1.8.7 doesn't have a `[]=` method. If you change it to the following, it will work:

```ruby
tab_a.versions = { "stable" => f.version }
```